File: day03/puzzle1.rb
#!/usr/bin/env ruby

# Login to https://adventofcode.com/2024/day/3/input to download 'input.txt'.

# lines = readlines
# lines = File.readlines("sample1.txt", chomp: true) # Answer: 161 (in 51 ms)
lines = File.readlines("input.txt", chomp: true) # Answer: 160672468 (in 59 ms)

REGEX = /mul\((\d{1,3}),(\d{1,3})\)/

multiplications = lines.map do |line|
  line.scan(REGEX).map do |n1, n2|
    n1.to_i * n2.to_i
  end
end

total = multiplications.flatten.sum

puts "total: #{total}"

File: day03/puzzle1.js
#!/usr/bin/env node

import fs from 'node:fs'

// fs.readFile('sample1.txt', 'utf8', (err, data) => {
fs.readFile('input.txt', 'utf8', (err, data) => {
    if (err) {
        console.error(err)
        return
    }

    const lines = data.split('\n').filter(s => s !== '')
    console.log('lines')
    console.log(lines)
    console.log()

    const regex = /mul\((?<n1>\d{1,3}),(?<n2>\d{1,3})\)/g
    const multiplications = lines.map(line => [...line.matchAll(regex)].map(({groups: {n1, n2}}) => n1 * n2)).flat()
    console.log('multiplications:')
    console.log(multiplications)
    console.log()

    const total = multiplications.reduce((acc, term) => acc + term, 0)
    console.log(`Total: ${total}`)
})
